src/mesh/interpolation.cxx

```diff
@@ -63,161 +64,179 @@ const Field3D interp_to(const Field3D &var, CELL_LOC loc)
-		// Producing a stencil centred around a lower Y value
-		s.pp = s.p;
-		s.p  = s.c;
-	      } else if(location == CELL_YLOW) {
-		// Stencil centred around a cell centre
-		s.mm = s.m;
-		s.m  = s.c;
-	      }
-	      
-	      result[i] = interp(s);
-	    }
-	  }
-	}
-	break;
-	// Need to communicate in Y
+              // Set stencils
+              s.c = var_fa[i];
+              s.p = var_fa[i.yp()];
+              s.m = var_fa[i.ym()];
+
+              if ((location == CELL_CENTRE) && (loc == CELL_YLOW)) {
+                // Producing a stencil centred around a lower Y value
+                s.pp = s.p;
+                s.p  = s.c;
+              } else if(location == CELL_YLOW) {
+                // Stencil centred around a cell centre
+                s.mm = s.m;
+                s.m  = s.c;
+              } else {
+                // We should never be here
+                throw BoutException("Please report this bug!");
+              }
+
+              result[i] = interp(s);
+            }
+          }
+        }
+        break;
+        // Need to communicate in Y
       }
       case CELL_ZLOW: {
         for(const auto &i : result.region(RGN_NOZ)) {
-	  s.c = var[i];
-	  s.p = var[i.zp()];
-	  s.m = var[i.zm()];
-	  s.pp = var[i.offset(0,0,2)];
-	  s.mm = var[i.offset(0,0,-2)];
-	  
-	  result[i] = interp(s);
-	}
-	break;
+          s.c = var[i];
+          s.p = var[i.zp()];
+          s.m = var[i.zm()];
+          s.pp = var[i.offset(0,0,2)];
+          s.mm = var[i.offset(0,0,-2)];
+
+          if ((location == CELL_CENTRE) && (loc == CELL_ZLOW)) {
+            // Producing a stencil centred around a lower Z value
+            s.pp = s.p;
+            s.p  = s.c;
+          } else if(location == CELL_ZLOW) {
+            // Stencil centred around a cell centre
+            s.mm = s.m;
+            s.m  = s.c;
+          } else {
+            // We should never be here
+            throw BoutException("Please report this bug!");
+          }
+
+          result[i] = interp(s);
+        }
+        break;
       }
       default: {
-	// This should never happen
-	throw BoutException("Don't know what to do");
+        // This should never happen
+        throw BoutException("Unsupported method of interpolation\n"
+                            " - don't know how to interpolate to %s",strLocation(loc));
       }
       };
-      
-      if(dir != CELL_ZLOW) {
-	// COMMUNICATION
-	
-	mesh->communicate(result);
-
-	// BOUNDARIES
 
+      if(dir != CELL_ZLOW) {
+        mesh->communicate(result);
       }
 
     }else {
       // Shifted -> shifted
-      // For now, shift to centre then to shifted
-      
+      // For now, shift to centre then to final location loc
+
       result = interp_to( interp_to(var, CELL_CENTRE) , loc);
     }
     result.setLocation(loc);
 
     return result;
   }
-  
+  if ((loc != CELL_CENTRE && loc != CELL_DEFAULT) && (mesh->StaggerGrids == false)) {
```

Comment:
Move this to be at the top of the function. It's better to put this kinds of input argument validation early on, so the reader can see what assumptions the function is making, what is invalid input, etc.